test/functional/test_framework/test_framework.py

```diff
@@ -155,14 +155,14 @@ def main(self):
             exit_code = self.shutdown()
             sys.exit(exit_code)
 
-    def parse_args(self):
+    def parse_args(self, test_file):
         previous_releases_path = os.getenv("PREVIOUS_RELEASES_DIR") or os.getcwd() + "/releases"
         parser = argparse.ArgumentParser(usage="%(prog)s [options]")
         parser.add_argument("--nocleanup", dest="nocleanup", default=False, action="store_true",
                             help="Leave bitcoinds and test.* datadir on exit or error")
         parser.add_argument("--noshutdown", dest="noshutdown", default=False, action="store_true",
                             help="Don't stop bitcoinds after the test execution")
-        parser.add_argument("--cachedir", dest="cachedir", default=os.path.abspath(os.path.dirname(os.path.realpath(__file__)) + "/../../cache"),
+        parser.add_argument("--cachedir", dest="cachedir", default=os.path.abspath(os.path.dirname(test_file) + "/../../cache"),
```

Comment:
> Is this correct? 

It is (mostly; see the last paragraph of this comment).

> now it is `test/functional/../../cache` (aka `cache`).

That is correct with a clarification. It is `build_dir/cache`, not `source_dir/cache`.

> Why not:
>
>```suggestion
>         parser.add_argument("--cachedir", dest="cachedir", default=os.path.abspath(os.path.dirname(__file__) + "/../../cache"),
>```

Using `__file__` won't work because it ends in the source tree on POSIX systems. For more details, please refer to https://discuss.python.org/t/symlink-resolution-in-starting-script-path/36537.

Perhaps, one path level can be skipped:
```suggestion
        parser.add_argument("--cachedir", dest="cachedir", default=os.path.abspath(os.path.dirname(test_file) + "/../cache"),
```
which ends in the `build_dir/test/cache` path.

Reply:
> > Perhaps, one path level can be skipped:
> 
> Yes, this is what I was trying to suggest. Previously the cache was in `test/cache` and now it should be as well, no?

Thanks! Done.